feat(android): getNotificationSettings support

vincent-paing opened the pull request:
Address https://github.com/invertase/notifee/issues/235

on Android, `getNotificationSettings` will return `alert` as `0` for permission denied and `1` for permission granted. This allows the library consumer to easily check if app level notification is enabled for Android on demand instead of settings up `onBackgroundEvent`. To preserve existing functionality, it will still returns `1` for the rest of the properties.

I added a unit test for the `getNotifcationSettings` for Android. However, I couldn't verify it works because I couldn't run the example app on my local machine with changes I made since it always point to the notifee github page for its dep. I will be more than happy to test it on emulator/simulator if someone could point me in the right direction.

Feel free to suggest fixes to the PR as well.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/notifee?pullRequest=237) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/notifee?pullRequest=237) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/notifee?pullRequest=237) it.</sub>

mikehardy reviewed (approved):
:wave:  I lived in Chiang Mai for half a year with my family and have really fond memories of Thailand :thailand: 

Thanks for posting this, it looks really clean - I can't see anything wrong actually

With regards to testing or demonstration I typically override the install in package.json to point to the correct directory, e.g. for the example I would point this to `../../` and yarn install again

https://github.com/invertase/notifee/blob/4a720dc8ac52506b302f6d9252a99cefc1d8cb28/packages/react-native/example/package.json?#L14

Similar for the tests_react_native e2e app - follow this to get the core lib right https://github.com/invertase/notifee/blob/main/CONTRIBUTING.md#step-2-install-test-project-dependencies and then in combo with a `../` package.json reference, you should be able to get the code in there to verify it?

The developer experience in the tests / example apps is pretty uncomfortable at the moment, apologies for that

I don't see a reason to block on that but it would be great to see a test in there if you already have one coded up - even if it were failing in CI right now (because the committed package.json files are referencing github#main we could  verify it locally and get them in and sort out developer experience after

mikehardy commented on packages/react-native/src/types/Module.ts:
```diff
@@ -426,7 +426,8 @@ export interface Module {
+   * The rest of the properties on the `IOSNotificationSettings` interface response return
    * as `AUTHORIZED`.
    *
    * @platform ios
```
no longer just ios - same for one below it on line 461/463

vincent-paing replied:
I assume I will have to delete the `@platform ios` line, correct?

mikehardy replied:
exactly that! nice

mikehardy commented:
Oh! One thing worth mentioning: docs - this should be mentioned in the docs I think? Maybe adding a page on the android side similar to the ios one here https://github.com/invertase/notifee/blob/main/docs/react-native/docs/ios/permissions.md

vincent-paing commented:
Thanks you for taking a look. I just recently moved to Thailand so still getting used to it here. I quite like it here as well! 🥰

It's currently 1 AM here, so I will take a look into it tomorrow morning and will follow your direction in setting up the local environment. After I have confirmed the behavior, will update the docs 👨‍💻

helenaford commented on packages/react-native/src/NotifeeApiModule.ts:
```diff
@@ -462,22 +462,21 @@ export default class NotifeeApiModule extends NotifeeNativeModule implements Mod
-        inAppNotificationSettings: 1,
-        authorizationStatus: 1,
-      } as IOSNotificationSettings);
+      return this.native.getNotificationSettings().then((alert: object) => {
```
```suggestion
      return this.native.getNotificationSettings().then(({ alert }: { alert: number }) => {
      
      ...
      return {
        alert,
        ...
        }
```

helenaford commented:
thanks for this addition! just had one thing... we also have a ticket to rename [ IOSNotificationSettings](https://github.com/invertase/notifee/issues/168) to `NotificationSettings` but that's probably best as a follow-up p/r as it will introduce breaking changes

mikehardy commented:
@helenaford I think this would be breaking - I just kicked off CI on the current iteration and it's running now.
Thinking larger scale I think we have a collection of "notification related savings":

- what are the app-wide settings for notifications for the app, ios and android?
- what is power saving enabled (including the special huawei power saving :roll_eyes: ) for the app?
- what is the status of the new Android12 set exact timer permission?
- for android what is the status of any particular channel?

And all of these things could change dynamically / asynchronously from anything else the app is doing

In my experience on the netinfo repo it was quite successful to have a "fetch state" API and a "listen to state" hook, each of which returned a typed state object. It's been a durable / maintainable interface https://github.com/react-native-netinfo/react-native-netinfo#usage

Over time the result value changes a little bit ("hey, we've got a new Android 12 thing you should know, it's a new property on the object) but people's code is forward-compatible and they don't have to change their API consumption patterns.

The object could look like

`fetchNotificationStatus`:

```
{
  'deviceInfo': PowerManagerInfo | undefined,
  'appPermission': {
    'notificationsEnabled': boolean,
    'alarmsEnabled': boolean,
    'isBatteryOptimizationEnabled': boolean,
  },
  'notificationSettings': {
    'general': {
      // ... all of the iOS settings here, badge, alert, etc
    },
    'channelGroups': [
       // one entry per android channel group, with a default group for ungrouped channels, e.g.:
      'defaultGroup': [
        showing channel status, e.g.:
        'excitingPromotions': AndroidChannel
      ],
      // ... rest of AndroidChannelGroups ?
    ]
  }
}
```

it still probably makes sense to fire events with detail information that represents what actually changed, but having a general shape defined for an "all of the state" call would help users I think so they could treat Notifee as the "store" instead of having to duplicate and manage all the state. Additionally a lot of these changes force-restart the app as a system behavior when permissions change, so the app will reboot and immediately need all the state to behave well.

Thoughts :thinking:  ?

vincent-paing commented:
> Edit: actually I think this will cause a breaking change on Android, as right now people will assume it will always return 1, so maybe it's worth it renaming in this p/r? But appreciate this will be asking more of your time! So happy to do it afterwards. 😎
 
 @helenaford I can include that as part of this PR if we're okay with the breaking changes. If I could share my suggestion, I like the way [react-native-permissions](https://github.com/zoontek/react-native-permissions) set up the structure. We can have a similar structure as part of refactoring the `iOSNotificationSettings`.

```ts
type NotificationStatus = "granted" | "blocked" | "provisional"

type NotificationSettings{
  status: NotificationStatus
  iOSSettings: iOSNotificationSettings
}
```
We can consume this API easily for just checking the permission and don't need to set default values for iOS only settings on Android platform.

mikehardy commented:
I had forgotten that it's not just a  boolean even, it is now an enum for ios notification permission. This stuff is *so* detail oriented it's crazy. Good catch on that. I also like react-native-permissions as a reference for API design here, it's a pleasant API to interact with

helenaford commented:
@mikehardy @vincent-paing I like both approaches. For simplicity, to begin with, we could start off with this - modifying our existing endpoint `getNotificationSettings`:
```
type NotificationSettings{
  status: NotificationStatus
  iOSSettings: iOSNotificationSettings
}  
```


And then as phase two, where we implement `fetchNotificationStatus`.

vincent-paing commented:
Sure, I will update the commit to this API tomorrow.

```ts
type NotificationSettings{
  status: AuthorizationStatus (existing type)
  iOSSettings: iOSNotificationSettings
}  
```

vincent-paing commented:
@helenaford Sorry it took me a while, was busy with new year trip :airplane:. 

Anyhow, things to note with the changes are 
- I couldn't find existing `AuthorizationStatus`  type so I went ahead and use `IOSAuthorizationStatus` instead. 
- It cause a duplication of `IOSAuthorizationStatus` since it's also inside `IOSNotificationSettings` type. I didn't want to remove it in `IOSNotificationSettings` because it might make bigger breaking changes and it impacts other areas. 
- I went ahead and edit the native module of Android to also return `authorizationStatus` instead of `alert` to be more consistent.
- I still couldn't get E2E test up and running so I haven't added test for that yet :cry:

helenaford commented:
@vincent-paing no problem at all. hope you had a great trip 😄 

- ah oops, yeah that's the type! I think I wrote `AuthorizationStatus` instead of `IOSAuthorizationStatus ` because we want to rename it to that.
- mmm, interesting re:e2e. are you getting errors when running the e2e commands? Maybe we need to update our README or fix something

vincent-paing commented:
I am able to run the e2e on my Macbook now, do you think this PR needs an end to end test?

mikehardy commented:
Well, I always like more testing vs less testing, but I also like more contributions then fewer contributions, the way those preferences interact is that I will certainly cheer for more tests :-), but I won't block on them. I'd merge and move forward if you didn't have time to add a test, personally. So then I think the final say goes to @helenaford. Cheers @vincent-paing

vincent-paing commented:
Since the test for notification settings requires interacting with System settings (needing to go to App Info to toggle notification on/off), I used [UIAutomator](https://developer.android.com/training/testing/other-components/ui-automator) to do this. However, I'm not sure how to link that test to `cavy`. Currently you would need to run `./gradlew androidTest` to execute this automation test.

I'm not sure how to do the same for iOS since I'm unfamiliar with iOS APIs.

Let me know if you have better approaches as well. This was the solution I ended up with.

vincent-paing commented:
@helenaford can you take a look at the PR and merge if it looks good to you?

mikehardy commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+---
+title: Permissions
+description: Android device grants permission for notification by default
```
This does not seem like a good description, when the page is about android permissions in general

```suggestion
description: Understanding and Checking Android notification permissions
```

mikehardy commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+  const channel = await notifee.getChannel();
+
+  if (channel.blocked) {
+    console.log('Channel notification has been disabled');
+  } else {
+    console.log('Channel notification has been enabled');
+  }
```
If the attribute is called "blocked", we should really decide on two words to represent the two states and use them everywhere.

I consulted the APIs and they appear to use "enabled" and "blocked". So in all cases I would propose we use those words if possible - attribute names, API / method names, documentation here

vincent-paing replied:
We will be using `enabled` for allowed state, and `blocked` for not-allowed state, correct?

mikehardy commented on packages/react-native/src/NotifeeApiModule.ts:
```diff
@@ -535,24 +536,28 @@ export default class NotifeeApiModule extends NotifeeNativeModule implements Mod
+        .then(({ authorizationStatus }: Pick<NotificationSettings, 'authorizationStatus'>) => {
+          return {
+            authorizationStatus,
+            iOSSettings: {
```
It seems strange that we are in an android conditional, and returning an iOSSettings object?

helenaford replied:
it was to avoid platform dependent code 🤔 But, maybe it's not needed as we have `authorizationStatus`?

vincent-paing replied:
Should it returns `undefined` instead of default values?

mikehardy commented on packages/react-native/src/types/Module.ts:
```diff
@@ -454,13 +449,10 @@ export interface Module {
-   * as `AUTHORIZED`.
-   *
-   * @platform ios
+   * On Android, `authorizationStatus` will return only either `IOSAuthorizationStatus.DENIED` or `IOSAuthorizationStatus.AUTHORIZED`
```
same issue here - seems strange that in a new world where we have Android status as well, we are tied to an IOS-specific object

helenaford replied:
yes, I believe this should be renamed to `AuthorizationStatus`

vincent-paing replied:
@helenaford should we introduce name updates as part of another PR?

mikehardy reviewed (changes requested):
This seems like it will all work - all of my comments were either
1) minor wording changes for documentation or, more importantly:
2) a real struggle for consistent, meaningful naming - I think that's actually easy to get right if we agree on a breaking change, it is just a refactor so the names are consistent and platform-neutral, but it requires a real decision to just do it @helenaford  ?

helenaford reviewed (commented):
@vincent-paing thanks so much for this pr 🙌 love that you added native tests too. Just added some initial comments, I will do another review tomorrow where I'll pull it down locally and run the changes on my devices. This is going to be an awesome addition to `notifee`

helenaford commented:
@mikehardy yes we want to rename `IOSAuthorizationStatus ` to `AuthorizationStatus` as it's impossible to avoid a breaking change, we want to do everything in one go.

I will get back to you about the `iOSSettings`... need to make sure that makes sense if we rename it to `settings`

vincent-paing commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+
+## Application wide
+
+To check whether the user has enabled application-wide notifications, call [`getNotificationSettings`](/reference/getnotificationsettings). The `authorizationStatus` attributes will returns `0` if user has disabled the permission, and `1` if it's granted.
```
Agreed that it should be an enum, we can use `ALLOWED`, `BLOCKED` ,`PROVISIONAL`.   For `PROVISIONAL`, it will only be for iOS?

vincent-paing commented on packages/react-native/src/types/Module.ts:
```diff
@@ -423,16 +419,15 @@ export interface Module {
    *
-   * On Android, all of the properties on the `IOSNotificationSettings` interface response return
-   * as `AUTHORIZED`.
+   * On Android, `authorizationStatus` will return only either `IOSAuthorizationStatus.DENIED` or `IOSAuthorizationStatus.AUTHORIZED`
```
I like`denied` and`authorized` better than`enabled` and `blocked`.

mikehardy replied:
Let's consolidate the naming choice here (I scattered it in a couple places, sorry) - @helenaford I am fine with "denied" and "authorized" (and, "provisional" I guess, for iOS). If you like that we have consensus, if not - I'll go with whatever you like. My main desire was for it to be consistent whatever the choice was :-)

vincent-paing commented:
@helenaford Let me know when you have update on the breaking naming changes.

vincent-paing commented on packages/react-native/src/types/Module.ts:
````diff
@@ -423,16 +419,15 @@ export interface Module {
    * ```
    *
-   * Use the other `IOSNotificationSettings` properties to view which specific permissions were
+   * For iOS specific settings, use the `IOSNotificationSettings` properties to view which specific permissions were
````
Just realized it has different key, I was using the name of the type  🤦

vincent-paing commented:
I went ahead and  changed `IOSAuthorizationStatus` to `IOSAuthorizationStatus`. I also moved that enum out of `NotificationiOS.ts` to `Notification.ts` since it no longer  depends on  platform. 

If you don't want to introduce breaking changes as part of this PR, let me know,  I can revert the commit. @helenaford

mikehardy reviewed (approved):
I'm okay with breaking changes personally I just try to be really good about helping people cope, specifically by putting a `!` on the commit line (like: feat(android)!: getNotificationSettings support` and adding a block at the end like:

BREAKING CHANGE: IOSNotificationStatus is renamed to NotificationStatus, and NotificationSettingsIOS is renamed to NotificationSettings now that android also has notification settings support"   (**please note I typed that out from memory without double-checking, might not be 100 but you get the idea**)

When I do that, I do not get any new issues in gihtub or many complaints so I think it works :-)

helenaford commented:
@vincent-paing @mikehardy yes, breaking changes are unavoidable even without the rename as we know can be returning false on Android where previously it would always be true. So yes happy with having all these changes here 😎 Will do one more review today, but looking good so far. Thanks for all your work on this @vincent-paing 🙌

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+
+On Android, notification permission is granted by default. However, a user may revoke notification permissions later through various means. In addition, a user may manage or revoke this notification permission at any of three levels; application-wide, channel groups and channels. 
+
+# Checking permission
```
```suggestion
# Checking permissions
```

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+
+## Application wide
+
+To check whether the user has enabled application-wide notifications, call [`getNotificationSettings`](/reference/getnotificationsettings). The `authorizationStatus` attributes will returns `DENIED` if user has denied the permission, and `AUTHORIZED` if it's granted.
```
```suggestion
To check whether the user has enabled notifications for your app, call [`getNotificationSettings`](/reference/getnotificationsettings). The `authorizationStatus` attributes will returns `DENIED` if user has denied the permission, and `AUTHORIZED` if it's granted.
```

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+  const channel = await notifee.getChannel();
+
+  if (channel.blocked) {
+    console.log('Channel notification has been authorized');
```
```suggestion
    console.log('Channel is denied');
```

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+  if (channel.blocked) {
+    console.log('Channel notification has been authorized');
+  } else {
+    console.log('Channel notification has been denied');
```
```suggestion
    console.log('Channel is authorized');
```

helenaford commented on docs/react-native/docs/android/permissions.md:
````diff
@@ -0,0 +1,59 @@
+}
+```
+
+For more understanding on channels, you can refer to [Channels & Groups](react-native/docs/android/channels).
````
```suggestion
> For more understanding on channels, you can refer to the [Channels & Groups](react-native/docs/android/channels) documentation.
```

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+
+## Permission events
+
+For listening to permission change events, refer to [Listening to channel events](react-native/docs/android/channels#listening-to-channel-events).
```
```suggestion
For listening to permission change events, refer to the [Listening to channel events](react-native/docs/android/channels#listening-to-channel-events) documentation.
```

helenaford commented on docs/react-native/docs/android/permissions.md:
```diff
@@ -0,0 +1,59 @@
+
+# Understanding permissions
+
+On Android, notification permission is granted by default. However, a user may revoke notification permissions later through various means. In addition, a user may manage or revoke this notification permission at any of three levels; application-wide, channel groups and channels. 
```
```suggestion
On Android, notification permissions are granted by default. However, a user may revoke them later through various means. In addition, a user may manage or revoke permissions at any of three levels; application-wide, channel groups and channels. 
```

vincent-paing commented:
@helenaford not sure why e2e android is failing, we didn't make any changes to  cancel API. I run it on my local machine and test passes. Is it a flaky test?

mikehardy commented:
@vincent-paing yes android e2e is a flaky test in CI at the moment - android e2e tests in general are really tough to stabilize in a CI environment, I'm battling with it across a few repos at the moment

vincent-paing commented:
@mikehardy Glad to know I'm not alone, Android E2E is a flaky mess in our react native project too! 😢 We're trying to fix it as much as possible but we still see flakiness from time to ime

helenaford commented:
@vincent-paing what type is `requestPermission` returning? I think we might have to change it for iOS... is it supposed to return NotificationSettings with a nested iosSettings?

vincent-paing commented:
@helenaford  It's returning  with nested `iOSSettings` afaik, Both `getNotificationSetting` and `requestPermission` returns same `NotificationSettings` type. Should it returns only `authorizationStatus` for Android, and `authorizationStatus` +`iOSSettings` for iOS?

helenaford commented:
@vincent-paing thanks, makes sense, no need to change Android. But I can see an issue with the iOS. I will push up a fix and it should be ready to merge to master. On iOS the native module isn't returning a nested `iOSSettings` object.

vincent-paing commented:
Good catch, I forgot to update the `requestPermission` iOS native code when  I change the type definition. Let me know if you need any help with it.

arcollector commented:
hello, what version did start using AuthorizationStatus instead of IOSAuthorizationStatus ?
